src/perf_counters.cc

```diff
@@ -0,0 +1,108 @@
+#include "perf_counters.h"
+
+#if defined HAS_LIBPFM
+#include <cstring>
+
+#include "pfmlib.h"
+#include "pfmlib_perf_event.h"
+#endif
+
+namespace benchmark {
+namespace internal {
+
+#if defined HAS_LIBPFM
+const bool PerfCounters::kSupported = true;
+
+bool PerfCounters::Initialize() { return pfm_initialize() == PFM_SUCCESS; }
+
+PerfCounters PerfCounters::Create(
+    const std::vector<std::string>& counter_names) {
+  if (counter_names.empty()) {
+    return NoCounters();
+  }
+  if (counter_names.size() > PerfCounterValues::kMaxCounters) {
+    std::cerr << counter_names.size()
+              << " counters were requested. The minimum is 1, the maximum is "
+              << PerfCounterValues::kMaxCounters << "\n";
+    return NoCounters();
+  }
+  std::vector<int> counter_ids(counter_names.size());
+
+  const int mode = PFM_PLM3;  // user mode only
+  for (size_t i = 0; i < counter_names.size(); ++i) {
+    const bool is_first = i == 0;
+    struct perf_event_attr attr;
+    std::memset(&attr, 0, sizeof(perf_event_attr));
+    attr.size = sizeof(attr);
+    const int group_id = !is_first ? counter_ids[0] : -1;
+    const auto& name = counter_names[i];
+    if (name.empty()) {
```

Comment:
you might want to use CHECK here too if it's not too catastrophic

Reply:
I'll leave it to the user of the object to decide what to do with the outcome, so I CHECK in benchmark-runner.cc